File: ImperatorToCK3/Source/Mappers/CoaMapper/CoaMapper.cpp
#include "CoaMapper.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "OSCompatibilityLayer.h"

mappers::CoaMapper::CoaMapper(const Configuration& theConfiguration)
{
	const auto coasPath = theConfiguration.getImperatorPath() + "/game/common/coat_of_arms/coat_of_arms";
	auto filenames = commonItems::GetAllFilesInFolderRecursive(coasPath);
	LOG(LogLevel::Info) << "-> Parsing CoAs.";
	registerKeys();
	for (const auto& fileName : filenames)
	{
		parseFile(coasPath + "/" + fileName);
	}
	clearRegisteredKeywords();
	LOG(LogLevel::Info) << "<> Loaded " << coasMap.size() << " coats of arms.";
}

mappers::CoaMapper::CoaMapper(const std::string& coaFilePath)
{
	registerKeys();
	parseFile(coaFilePath);
	clearRegisteredKeywords();
}

void mappers::CoaMapper::registerKeys()
{
	registerKeyword("template", commonItems::ignoreItem); // we don't need templates, we need CoAs!
	registerRegex(commonItems::catchallRegex, [this](const std::string& flagName, std::istream& theStream) { // the rest should be CoAs
		auto coaItem = commonItems::stringOfItem(theStream).getString();
		coasMap.insert(std::pair(flagName, coaItem));
	});
}

std::optional<std::string> mappers::CoaMapper::getCoaForFlagName(const std::string& impFlagName)
{
	if (coasMap.contains(impFlagName)) return coasMap.at(impFlagName);
	return std::nullopt;
}

File: ImperatorToCK3/Mappers/CoaMapper.cs
using System.Collections.Generic;
using commonItems;
using System.IO;

namespace ImperatorToCK3.Mappers {
    public class CoaMapper : Parser {
        public CoaMapper(Configuration theConfiguration) {
            var coasPath = Path.Combine(theConfiguration.ImperatorPath, "game", "common", "coat_of_arms", "coat_of_arms");
            var fileNames = SystemUtils.GetAllFilesInFolderRecursive(coasPath);
            Logger.Log(LogLevel.Info, "Parsing CoAs.");
            RegisterKeys();
            foreach (var fileName in fileNames) {
                ParseFile(Path.Combine(coasPath, fileName));
            }
            ClearRegisteredRules();
            Logger.Log(LogLevel.Info, "Loaded " + coasMap.Count + " coats of arms.");
        }
        public CoaMapper(string coaFilePath) {
            RegisterKeys();
            ParseFile(coaFilePath);
            ClearRegisteredRules();
        }
        private void RegisterKeys() {
            RegisterKeyword("template", ParserHelpers.IgnoreItem); // we don't need templates, we need CoAs!
            RegisterRegex(CommonRegexes.Catchall, (reader, flagName) => {
                coasMap.Add(flagName, new StringOfItem(reader).String);
            });
        }

        public string? GetCoaForFlagName(string impFlagName) {
            bool contains = coasMap.TryGetValue(impFlagName, out string? value);
            return contains ? value : null;
        }

        private Dictionary<string, string> coasMap = new();
    }
}
